### PR TITLE
use versions range to search for duplicated functions

### PR DESCRIPTION
**Pull request by isfedorov**
Find duplicated functions based on versions intersection like
```
#[PhpStormStubsElementAvailable(from: '7.1', to: '7.4')]
function bcmod (string $num1, string $num2, ?int $scale = 0)
{}
#[PhpStormStubsElementAvailable(from: '7.2', to: '8.0')]
function bcmod (string $num1, string $num2, ?int $scale = 0)
{}
```

### CODE REVIEW

File: tests/StubsTest.php [hunk 1]
```diff
@@ -377,15 +379,50 @@ private static function getDuplicatedFunctions(array $filtered): array
...
-                $sinceVersionOfInitialFunction = Utils::getDeclaredSinceVersion(
+                $functionVersions[] = Utils::getAvailableInVersions(
                     PhpStormStubsSingleton::getPhpStormStubs()->getFunctions()[$value->name]);
-                $sinceVersionsOfDuplicatedFunctions = array_map(function (PHPFunction $function) {
-                    return Utils::getDeclaredSinceVersion($function);
+                $functionVersions[] = array_map(function (PHPFunction $function) {
+                    return Utils::getAvailableInVersions($function);
                 }, $duplicatesOfFunction);
-                return !empty(array_search($sinceVersionOfInitialFunction, $sinceVersionsOfDuplicatedFunctions));
+                $twoDimensionalArray = [];
```
Comment (MaXal) [inline tests/StubsTest.php:387]: This part is really hard to follow. Do we have the following structure here: `["php_version" => ["function1", "function2"]`?

**Review by MaXal — CHANGES_REQUESTED** [review]

File: tests/Model/BasePHPElement.php [hunk 2]
```diff
@@ -113,25 +115,34 @@ protected static function findTypeFromAttribute(array $attrGroups): ?string
...
+                        $arg = $attr->args[0]->value;
+                        if ($arg instanceof Array_) {
+                            $value = $arg->items[0]->value;
+                            if ($value instanceof String_) {
+                                return ['from' => $value->value];
+                            }
+                        } else {
+                            $rangeName = $attr->args[0]->name;
+                            return $rangeName === null || $rangeName->name == 'from' ?
+                                ['from' => $arg->value, 'to' => '8.0'] :
```
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/Model/BasePHPElement.php:138]: What if function will be declared with attribute `#[PhpStormStubsElementAvailable(from: '7.1', to: '8.0')]`. How should we compare '8.0' and 'latest'?

File: tests/Parsers/Utils.php [hunk 3]
```diff
@@ -48,6 +46,37 @@ public static function getDeclaredSinceVersion(BasePHPElement $element): ?string
...
+            $latestVersions[] = self::getLatestAvailableVersionsFromParentClass($element);
+        }
+        $flattenedArray = Utils::flattenArray($latestVersions, false);
+        sort($flattenedArray, SORT_DESC);
+        return array_pop($flattenedArray);
+    }
+
+    public static function getAvailableInVersions(BasePHPElement $element): array
+    {
+        $versions = [5.3, 5.4, 5.5, 5.6, 7.0, 7.1, 7.2, 7.3, 7.4, 8.0];
```
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/Parsers/Utils.php:68]: These versions are used to get range of php versions where function is available so it's not enough to have just first and last supported versions. To say the truth I don't see any problem with explicit versions declaration as even phpstorm has hardcoded language level versions.

File: tests/Parsers/Utils.php [hunk 4]
```diff
@@ -48,6 +46,37 @@ public static function getDeclaredSinceVersion(BasePHPElement $element): ?string
...
+        $firstSinceVersion = self::getDeclaredSinceVersion($element);
+        if ($firstSinceVersion === null) {
+            $firstSinceVersion = 5.3;
+        }
+        $lastAvailableVersion = self::getLatestAvailableVersion($element);
+        if ($lastAvailableVersion === null) {
+            $lastAvailableVersion = 8.0;
```
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/Parsers/Utils.php:75]: explicit version is required for comparing of declared versions...two duplicated functions can have attributes like `#[PhpStormStubsElementAvailable(from: '7.1', to: '8.0')]` and `#[PhpStormStubsElementAvailable(from: '7.1')]` so they are duplicates but one has latest version as '8.0' and other as 'latest'. How to compare them? We can defined some constant like `LATEST_PHP_VERSION = 8.0` but I'm not sure that will be really much better. This constant still will have to be updated with every new PHP version

File: tests/StubsTest.php [hunk 5]
```diff
@@ -377,15 +379,50 @@ private static function getDuplicatedFunctions(array $filtered): array
...
+            } else {
+                self::convertArrayToTwoDirectional($element, $resultArray);
+            }
+        });
+    }
+
+    #[Pure]
+    private static function getIntersection(array $next, array $twoDimensionalArray): array
+    {
+        $arg_list = func_get_args();
```
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/StubsTest.php:418]: Well yes..it's a bit tricky and this is why: suppose we have an original function `foo()` with declared version like `[7.1,7.2,7.3]` and this function has 2 duplicates with declared versions `[7.2,7.3,7.4]` and `[7.3,7.4,8.0]` so as result array of all available versions is `[[7.1,7.2,7.3],[7.2,7.3,7.4],[7.3,7.4,8.0]]` and we should get the intersection of all these arrays. The problem is that built-in `array_intersect` doesn't intersect properly all 3 arrays, only 2. That's why I've to manually unwrap array with `...` in function call and go through these arrays as separate parameters to intersect them one by one.

File: tests/StubsTest.php (hunk 1 above)
Comment (isfedorov) [reply tests/StubsTest.php:387]: Initially, we get an array with versions that can look like [[7.1,7.2],[[7.1]]] so we want to convert it to unified form for further processing and that's why `$twoDimensionalArray = [];` is initialized to later fill it with arrays of versions to get array like [[7.1,7.2],[7.1]]

File: tests/Model/BasePHPElement.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaXal) [reply tests/Model/BasePHPElement.php:138]: Let's introduce a class `PHPVersions` which will contain a list of all existing versions from `Utils.php` and let's take `$versions[-1]` instead of '8.0`. That will reduce the number of places to update to the single source of truth.

File: tests/Parsers/Utils.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaXal) [reply tests/Parsers/Utils.php:68]: See my comment above.

File: tests/Parsers/Utils.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaXal) [reply tests/Parsers/Utils.php:75]: See comment above.

File: tests/StubsTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaXal) [reply tests/StubsTest.php:418]: The intersection of your arrays is `7.3` (by definition intersection includes elements that exist in all arrays). And PHP works exactly like this: https://3v4l.org/relZX

Probably you're looking for something else? What is the expected result? `[7.2, 7.3, 7.4]`?

File: tests/StubsTest.php (hunk 1 above)
Comment (MaXal) [reply tests/StubsTest.php:387]: What is the place that outputs `[[7.1]]`? It looks like it shouldn't and this flattening is a workaround.

File: tests/StubsTest.php [hunk 6]
```diff
@@ -377,15 +379,50 @@ private static function getDuplicatedFunctions(array $filtered): array
...
-                $sinceVersionsOfDuplicatedFunctions = array_map(function (PHPFunction $function) {
-                    return Utils::getDeclaredSinceVersion($function);
+                $functionVersions[] = array_map(function (PHPFunction $function) {
+                    return Utils::getAvailableInVersions($function);
                 }, $duplicatesOfFunction);
-                return !empty(array_search($sinceVersionOfInitialFunction, $sinceVersionsOfDuplicatedFunctions));
+                $twoDimensionalArray = [];
+                self::convertArrayToTwoDirectional($functionVersions, $twoDimensionalArray);
+                $hasDuplicates = false;
+                while (($next = array_pop($twoDimensionalArray)) !== null && !empty($twoDimensionalArray)) {
```
Comment (MaXal) [inline tests/StubsTest.php:390]: Will the following work exactly the same?
```
$hasDuplicates = false;
$current = array_pop($array);
while ($next = array_pop($array)) {
    if (!empty(array_intersect($current, $next))) {
        $hasDuplicates = true;
    }
    $current = $next;
}
```

**Review by MaXal — CHANGES_REQUESTED** [review]

File: tests/StubsTest.php (hunk 1 above)
Comment (isfedorov) [reply tests/StubsTest.php:387]: Ok. So here is a simple example.
Screenshot 2020-12-29 at 23 17 52
But it's possible that there will be several duplicates 
Plain array comes from `Utils::getAvailableInVersions(PhpStormStubsSingleton::getPhpStormStubs()->getFunctions()[$value->name]);` and other from duplicates and it's required to convert them

File: tests/StubsTest.php (hunk 6 above)
Comment (isfedorov) [reply tests/StubsTest.php:390]: Seems this code will not work if we have an original function and 2 duplicates and want to find an intersection of the original (1st) function with 3rd function as code above finds intersection only of 2 adjacent elements but it's required to jump over middle elements. like
`#[PhpStormStubsElementAvailable(from: '5.6', to: '7.3')]`
`#[PhpStormStubsElementAvailable(from: '7.4', to: '8.0')]`
`#[PhpStormStubsElementAvailable(from: '7.2', to: '7.3')]`
In such cases first and the last functions would be duplicates.

File: tests/StubsTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/StubsTest.php:418]: >And PHP works exactly like this
And this is actually the problem :) it's required to find not only intersection of elements from first array in other arrays but also find intersection of elements from second array in third array etc.

File: tests/Model/BasePHPElement.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (isfedorov) [reply tests/Model/BasePHPElement.php:138]: Yes, this is a good solution! Thank you for the suggestion. I'll fix this.

File: tests/StubsTest.php (hunk 6 above)
Comment (MaXal) [reply tests/StubsTest.php:390]: Good catch! It should be:
```
$hasDuplicates = false;
$current = array_pop($array);
while ($next = array_pop($array)) {
    if (!empty(array_intersect($current, $next))) {
        $hasDuplicates = true;
    }
    $current = array_merge($current, $next);
}
```

File: tests/StubsTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaXal) [reply tests/StubsTest.php:418]: See my example below that do exactly what we need but using less magic and built-in functions.

File: tests/StubsTest.php (hunk 6 above)
Comment (isfedorov) [reply tests/StubsTest.php:390]: Hm, `array_merge` will do the trick I guess. Thank you for the suggestion!

File: tests/StubsTest.php (hunk 1 above)
Comment (MaXal) [reply tests/StubsTest.php:387]: Let's do the following instead than:
```
array_push($functionVersions, array_values(array_map(function (PHPFunction $function) {
                    return Utils::getAvailableInVersions($function);
}, $duplicatesOfFunction)));```
